Documentation for type checker prototype (#13350)
enso-org/enso: docs/types/inference-and-checking.md

checker on its own is quite useless. For Enso to truly be usable, it must also
have a powerful type inference engine.

<!-- MarkdownTOC levels="2,3" autolink="true" -->

- [Type Checker Prototype](#type-checker-prototype)
- [Design Goals](#design-goals)
  - [Maximal Inference Power](#maximal-inference-power)
  - [Type Inference Algorithm](#type-inference-algorithm)
  - [Type Checking Algorithm](#type-checking-algorithm)

<!-- /MarkdownTOC -->

> ## Type Checker Prototype

Reviewer: ```suggestion
## Type Checker Implementation
```

> A prototype of a type checker has been developed that does try to reach all the

Reviewer: > does try

or "does not try"?

design goals written below, but tries to perform best effort type checking to
provide lints and warnings to developers to aid in development.

> As it is a prototype and may slow down the compilation, it is only enabled if an

Reviewer: ```suggestion
As static type checking isn't needed for Enso execution, the type checker is only enabled if an
```

`--enable-static-analysis` flag has been passed to the compiler.

To try out the type checker you may run

```bash
./built-distribution/enso-engine-0.0.0-dev-windows-amd64/enso-0.0.0-dev/bin/enso --compile /path/to/Project --enable-static-analysis
```

### Overall Design

#### What is being checked

Any issues found by the type checker are reported as compilation warnings. They
are not treated as errors for two reasons: firstly, because the type checker is
still a prototype and so it should not block programs in case there is a mistake
in its logic, and secondly, because the program can still run - if the given
problematic block of code that is guaranteed to error is not reached during
execution it may not cause problems.

Currently the type checker reports the following kinds of warnings:

> - type mismatch, reported whenever during a method call the inferred type of an

Reviewer: ```suggestion
- **type mismatch**, reported whenever during a method call the inferred type of an
```

  argument passed to the method does not satisfy the type expected by the
  function's signature, e.g. calling function defined as
  `function (x : Integer) = ...` with an argument of type text:
  `function "Txt"`,

> - no such method, whenever a method is being called on an object with an

Reviewer: ```suggestion
- **no such method**, whenever a method is being called on an object with an
```

  inferred type and the type checker deduces that the inferred type does not
  have a method with the given name, e.g. when calling `x.method_1` and the type
  of `x` is not `Any` and does not define method called `method_1`,

> - discarded value, reported whenever a value, whose inferred type is a

Reviewer: ```suggestion
- **discarded value**, reported whenever a value, whose inferred type is a
```

The type inference algorithm tries to infer the types wherever it is possible,
but it is designed to give up in a graceful way. Static type errors are only
reported if the type checker can 'prove' that a given operation **will** surely
fail at runtime (if the piece of code is reached). If the operation may fail or
succeed, no errors are reported.

This makes `Any` a special type. In terms of the subtyping relationship it is a
[top type](https://en.wikipedia.org/wiki/Top_type), however in terms of the type
checker it is a bit closer to the
[bottom type](https://en.wikipedia.org/wiki/Bottom_type) - because a value of
type `Any` can be of any particular type, that means no error will be reported
because there is no way to guarantee a failure at runtime. Arguments of type
`Any` can be passed to methods expecting all kinds of types (then, they can fail
at runtime, but there is no way to prove a guaranteed failure statically), and
all methods can be called on `Any` (not only those defined on `Any`, as the
actual value passed in can have the given method defined). It behaves similarly
to
[`any` type in TypeScript](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#any)

> or [`dynamic` type in Kotlin](https://kotlinlang.org/docs/dynamic-type.html).

Reviewer: `dynamic` had also been [considered for Java](https://wiki.openjdk.org/pages/viewpage.action?pageId=7897244), but it was dropped from the plans.

Because of that, the `Any` type can be used to represent unexpected return
values from polyglot calls.

#### Relation to Intersection Types

> Our approach to intersection types that allows for a value to have multiple

Reviewer: A link to the "Enso approach" description might be handy.

different types, some of which are hidden, adds some challenges to the type
checking. The current state (where the hidden part of the type can only be
uncovered via an explicit cast) seems to strike a relatively good balance
between the ability to work with intersection types in the interactive GUI (that
is capable of inserting the necessary casts) and pass them through various
checked methods (which may hide the extra parts, but still being able to uncover
them) and the capabilities of static analysis of types.

Any changes to the related logic must be done very carefully as it is very easy
to introduce a change that will introduce a 'collapse' of the type system that
would make it unable to report any useful warnings.

> For example, if uncovering hidden values was possile when calling a method with

Reviewer 2: ```suggestion
For example, if uncovering hidden values was possible when calling a method with
```

checked arguments (without explicit casts), it would no longer be possible to
report any type mismatch warnings. That is because even if there's a value
`x : Integer` and function `f x` expects a `Text`, we cannot rule out that `x`
was not createad as `Integer & Text` before and the `Text` part was hidden. As
the hidden part of the type is never known _statically_, the static analysis
cannot rely on it and must assume that every value can have any kind of types in
the hidden part. Thus only relying on the ability to 'uncover' the hidden parts
via explicit syntactic constructs allows to keep the ability to reason about

> them in the static analysis (unless the system is constructed on some completely

Reviewer: Why doesn't the sentence talk about what currently works but rather mentions that _"no good alternatives have been discussed"_?

Author: That sentence was a kind of "I believe it's impossible but don't want to say impossible because I know I may be surprised".

I was describing the fact that we do not allow uncovering hidden parts in method calls. What currently works is simply keeping not allowing that but the sentence was hypothesizing other approaches where we **do allow** it somehow.

Not sure what I should change in this paragraph but happy to take suggestions.

Reviewer: - whether `:` statement succeeds or not is beyond the scope of static analysis
   - as such `:` statements never yield any static errors
   - they are only treated as an evidence (is that the right word?) that the returned value must have the requested type, if no exception was raised
- all other language constructs are subject of static analysis
   - e.g. method invocations
   - passing values as arguments
   - using them in binary operators
   - all of this can be checked statically in Enso

Author: Thanks, that's a nice summary. I've tried to port it to the docs in f79b564705687465c14f85eb352641dcfda56bda

different approach, but currently no good alternatives have been discussed).

#### Local Inference and Type Propagation

The type inference relies on existing type signatures and type assertions. Since
function argument types are checked at runtime, the type checker treats them as
assertions that an incoming value is of a given type. Similarly, code following
a type assertion inside of an expression (`y = x : T`, or `(x:T).method`) relies
on the fact that the control flow only proceeds if that assertion was satisfied.

> The processing is performed by traversing the IR of each method body bottom-up.

Reviewer: > method body bottom-up

- _"is the last statement in a block (e.g. **bottom**) processed first?"_
- _"and then previous ones statements **up** to the first statement?"_
- no, but that was the first thing I imagined when reading "bottom-up""
- maybe from _terminal IR elements to more complex ones_?

First types of the leafs - literals or variables - are inferred, and then based
on their types, the type of more complex 'nodes' (e.g. function application) is
derived based on very basic inference rules that stem from the
[simply typed lambda calculus](https://en.wikipedia.org/wiki/Simply_typed_lambda_calculus)
with some additional extensions for calling methods on objects and other
features of the language like pattern matching. Inside of blocks of code, the
inference is run line-by-line; whenever the assignment operator `=` creates a
new binding, the inferred type of that binding is saved in a mapping that can
then be used by subsequent lines of the block that can reference that binding.

Calling of methods on objects is implemented via special logic for application
on `UnresolvedSymbol`. That is because within the IR, the expression `a.f b c`

> is actually translated to `UnresolvedSymbol<f> a b c`.

Reviewer: +1 - quite a nice formulation of what I was only feeling (in debugger)

> The most important rules for inference are then: ...

Author: I was thinking of adding some descriptions for the most important propagation rules as implemented in `TypePropagation::tryInferringType` but from a more high-level overview perspective.

Do you think its worth adding?

Reviewer 2: Few senteces about it won't do any harm. Go ahead.

Reviewer: This _"Local Inference and Type Propagation"_ section is the section I was waiting for. It has enlighten me already.

> descriptions for the most important propagation rules

+1

These rules are implemented in `TypePropagation::tryInferringType` and its
related helper methods.

### Overall structure of the implementation

The implementation consists of three phases:

> 1. `TypeInferenceSignatures` analyzes the IR of method definitions and

Reviewer: What's the pass relation to `TypeSignatures`?

Author: Good question.

Honestly I don't fully remember.

`TypeInferenceSignatures` reads from `TypeSignatures`.

I think `TypeSignatures` was mostly associating the signatures for things like inline expressions `x = a:T` and then also return type signatures that were rewritten at some stage. The `TypeInferenceSignatures` was reading that to reconstruct function signatures.

But tbh I'm now wondering if the two could be merged. I must have seen `TypeSignatures` when writing `TypeInferenceSignatures` so I assume there was likely a reason why I couldn't just use results of `TypeSignatures` directly and needed this separate pass. But I really can't remember what it was. Or it is also possible that it was just a mistake.

Reviewer: > the two could be merged

OK, let that be a subject of some further work.

   associates a type signature with each method based on its type ascriptions.
   If a method has no type ascriptions, the argument types and the return type
   will default to unknown type (`Any`) but the method will still get a
   signature at least indicating the _arity_ of the method.
2. `StaticModuleScopeAnalysis` builds a static counterpart of `ModuleScope` that
   holds method definitions available in scopes of each module. The static
   module scope is then used to resolve methods on types during the static
   analysis, in the same way as `ModuleScope` is used at runtime for method
   dispatch. i. The `BuildScopeFromModuleAlgorithm` and
   `MethodResolutionAlgorithm` encapsulate the core logic, so that the same
   logic is guaranteed to be used both at runtime and compile-time.
3. `TypeInferencePropagation` which analyzes each method, tries to infer types
   of each sub-expression and report any issues found. While a type is inferred
   for every sub-expression inside of a method body, the types are stored in

>    metadata only for the named bindings to conserve memory. The rationale is

Reviewer: ```suggestion
   metadata **only for the named bindings** to conserve memory. The rationale is
```

    provided, or the `...` operator is used to explicitly stop the default
    application).
- checking for `Private_Access` violations statically.
  - Currently the `StaticModuleScopeAnalysis` pass does not record if a given
    method or constructor (or the whole module) is marked as `private`. This
    property should be recorded, and if a private method is called from an
    outside module, a warning about the private access could be reported also
    statically.

#### Integration with the VS Code extension

Integrating the type checker with the VS Code extension for Enso can lead to
vastly improved developer experience. The inferred types of bindings can be used
to offer some form of method autocomplete and warnings could be displayed inline
inside of the code editor.

#### More powerful inference

The current approach for inference is very simple, but can already provide basic
checking helpful during development. If types of method arguments are known and
method definitions on types have checked signatures, then most expressions can
be inferred and the type checker can provide warnings in case the inferred types

> do not match the expectations, allowing the developer to find bugs before before

Reviewer: ```suggestion
do not match the expectations, allowing the developer to find bugs before
```

running the program.

However, the inference algorithm has been created with simplicity and checking
for program correctness in mind, so the inference does not work with ambiguity.
If function arguments do not have checked types, they are treated as `Any` and
very little information can be derived from them. Currently, the algorithm does
not try to do bidirectional inference that would try to guess what the types of

> the function 'should be' based on how they are used.

Reviewer: - types of functions are either specified explicitly
- or they are set to `Any` - according to the arity of the function arguments

Possible future improvement is to give each argument a new type variable and check the result of unification.

- is there _any unification_ happening right now at all?
- what it would take to find out that `[ "a" ].at 0` is a `Text` (assuming we cannot deduce that right now)?

Author: I tried to answer the questions in 51e276beff773c99a576ca6e9c1be9abda0b32a3
